File: 2021_12_24_advent_of_code_in_fsharp/AdventOfCode/AdventOfCode2020/Day01_sum.fs
module AdventOfCode2020.Day01

open Common

open System


let loadData (input: string) = input.Split Environment.NewLine |> Seq.map Int32.Parse

let puzzle1 input =
    let (a, b) = loadData input |> allUniquePairs |> Seq.find (fun (a, b) -> a + b = 2020)
    (a * b).ToString()

let puzzle2 input =
    let (a, b, c) = loadData input |> Seq.toArray |> allUniqueTriples |> Seq.find (fun (a, b, c) -> a + b + c = 2020)
    (a * b * c).ToString()



// using general function 'allUniqueTuples' instead of 'allUniquePairs', 'allUniqueTriples'
let puzzle' input n =
    let result =
        loadData input |> Seq.toList |> allUniqueTuples n |> Seq.find (fun lst -> List.sum lst = 2020) |> Seq.reduce (*)
    result.ToString()

let puzzle1' input = puzzle' input 2
let puzzle2' input = puzzle' input 3



// improve efficiency by using map of sums left
let findSumOfN n target numbers =
    let rec loop items map lists =
        match items with
        | [] -> None
        | item :: rest ->
            match Map.tryFind item map with
            | Some lst -> Some(item :: lst)
            | None ->
                let completed, lists' = insertThenPartition (n - 1) item lists
                let map' =
                    completed
                    |> Seq.fold (fun m lst -> m |> Map.change (target - List.sum lst) (Option.orElse (Some lst))) map
                loop rest map' lists'
    loop numbers Map.empty []


let puzzle'' input n =
    let result = loadData input |> Seq.toList |> findSumOfN n 2020 |> Option.get |> Seq.reduce (*)
    result.ToString()

let puzzle1'' input = puzzle'' input 3
let puzzle2'' input = puzzle'' input 3

File: 2023_10_03_advent_of_code_in_kotlin/AdventOfCode/src/main/kotlin/adventOfCode2020/day01_sum.kt
package adventOfCode2020.day01_sum

import common.*

import common.allUniquePairs
import common.allUniqueTriples
import common.allUniqueTuples

fun loadData(input: String) = input.lineSequence().map { it.toInt() }

fun puzzle1(input: String): String {
    val (a, b) = loadData(input).allUniquePairs().first { (a, b) -> a + b == 2020 }
    return (a * b).toString()
}

fun puzzle2(input: String): String {
    val (a, b, c) = loadData(input).toList().allUniqueTriples().first { (a, b, c) -> a + b + c == 2020 }
    return (a * b * c).toString()
}


// for n=3 and triple no 787322 is found (217+196+1607=2020) and the execution takes 2m 1.872156193s :) in F# 1s
fun puzzle(input: String, n: Int): String {
    val result = loadData(input).allUniqueTuples(n).first { lst -> lst.toSequence().sum() == 2020 }
        .toSequence().reduce { a, b -> a * b }
    return result.toString()
}

fun puzzle1b(input: String) = puzzle(input, 2)
fun puzzle2b(input: String) = puzzle(input, 3)

